flight/PiOS/STM32F4xx/pios_dmashot.c

```diff
@@ -0,0 +1,423 @@
+#include <pios.h>
+#include "pios_dmashot.h"
+
+#define MAX_TIMERS					8
+
+// This is to do half-word writes where appropriate. Still needs
+// code for it and verification whether it does something.
+// Full word writes seem to work on 16bit timers, but for an
+// effort to save memory.
+union dma_buffer {
+	void *ptr;
+	uint32_t *fw;
+	uint16_t *hw;
+};
+
+struct servo_timer {
+	TIM_TypeDef *timer;									// What timer
+	uint32_t sysclock;									// Flava Flav!
+
+	const struct pios_dmashot_timer_cfg *dma;			// DMA config
+	const struct pios_tim_channel *servo_channels[4];	// Original channels
+
+	// Tracking used TIM channels to allocate least memory.
+
+	uint8_t low_channel;								// Lowest known TIM channel
+	uint8_t	high_channel;								// Duh
+
+	union dma_buffer buffer;							// DMA buffer
+
+	TIM_OCInitTypeDef *ocinit;
+};
+
+#define TIMC_TO_INDEX(c)			((c)>>2)
+#define NUM_CHANNELS(s)				(TIMC_TO_INDEX((s)->high_channel - (s)->low_channel) + 1)
+
+#define DMA_DISABLE(s)				{ DMA_Cmd(s, DISABLE); while (DMA_GetCmdStatus(s) == ENABLE); }
+#define DMA_ENABLE(s)				{ DMA_Cmd(s, ENABLE); /* while (DMA_GetCmdStatus(s) == DISABLE); */ }
+
+#define TIM_DISABLE(t)				{ TIM_Cmd(t, DISABLE); while (t->CR1 & TIM_CR1_CEN); }
+#define TIM_ENABLE(t)				{ TIM_Cmd(t, ENABLE); while (!(t->CR1 & TIM_CR1_CEN)); }
+
+#define ERROR()						PIOS_Assert(0)	// return -1
+
+const struct pios_dmashot_cfg *dmashot_cfg;
+struct servo_timer **servo_timers;
+
+int PIOS_DMAShot_GetFIFOCadence(int num_chan)
+{
+	switch(num_chan)
+	{
+		default:
+		case 1:
+			return DMA_FIFOThreshold_1QuarterFull;
+		case 2:
+			return DMA_FIFOThreshold_HalfFull;
+		case 3:
+			return DMA_FIFOThreshold_3QuartersFull;
+		case 4:
+			return DMA_FIFOThreshold_Full;
+	}
+}
+
+void PIOS_DMAShot_Init(const struct pios_dmashot_cfg *config)
+{
+	dmashot_cfg = config;
+}
+
+void PIOS_DMAShot_Prepare()
+{
+	if (dmashot_cfg && !servo_timers) {
+		// Allocate memory
+		servo_timers = PIOS_malloc_no_dma(sizeof(struct servo_timer) * MAX_TIMERS);
+		if(!servo_timers)
+			ERROR();
```

Comment:
This hurts readability IMO. Does it return, does it assert? Makes life harder when chasing bugs.